std/mem.zig

```diff
@@ -254,6 +255,27 @@ pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     return true;
 }
 
+/// Compares two slices and returns whether they are equal in constant time.
```

Comment:
It doesn't, does it? It returns T, not bool.

If you change the return type to bool, you run into the issue that a sufficiently smart compiler may short-circuit the loop.

I suppose that, in theory, that's already possible with zig's closed world compilation model. LLVM can look at the contexts in which a function is called and specialize when it infers a boolean context. I don't think it's actually smart enough right now to pull that off but that's shaky ground to be on.

Rather than `@noInlineCall`, what is probably needed here is an attribute that says "DWIM, don't optimize at all." Such an attribute doesn't exist at the moment but it'd be `noinline optnone` instead of just `noinline`.

Alternatively, maybe it's possible to convince llvm that all loads are volatile and cannot be elided. However, I'm not sure how strong llvm's guarantees are here. Does any kind of alias analysis weaken it?

FWIW, libraries like openssl implement constant-time operations in asm because, even though it's a bigger maintenance hassle, it's easier to reason about the run-time behavior.

Reply:
I actually think it might be worth having a compiler builtin for a constant time equality check. Trying to implement this in userland is fighting against a foundational premise that the rest of the codebase shares: that the goal of the compiler is to interpret the semantic meaning and then emit whatever machine code it can to make it go fast. Here the goal is different: make it go the same amount of time regardless of the input.

The benefit of a compiler builtin is that the compiler would understand how to do the operation at compile-time, as well as lower to whatever code is appropriate per-architecture, be it LLVM IR with `optnone` or assembly.

One thing I'd like to find out is whether there is a need for more constant time primitives. Looking at https://github.com/pornin/CTTK for inspiration, it seems they have API for all manner of integer operations. I wonder what operations crypto libraries depend on.